### PR TITLE
Add validation to date parser

### PR DESCRIPTION
**Pull request by HuwCampbell**

### CODE REVIEW

File: src/Icicle/Data/DateTime.hs [hunk 1]
```diff
@@ -31,11 +33,47 @@ renderDate
...
- $ D.DateTime y m d 0 0 0
+ | isValid y m d =
+   Right
+   $ DateTime
+   $ D.DateTime y m d 0 0 0
+ | otherwise =
+   Left $ "Invalid date: " <> show y <> "-" <> show m <> "-" <> show d
+
+isValid :: Int -> Int -> Int -> Bool
+isValid year month day =
```
Comment (amosr) [inline src/Icicle/Data/DateTime.hs]: oh, gosh. there's really nothing to do this in Data.Time or Data.Dates?

Comment (amosr) [inline src/Icicle/Data/DateTime.hs]: gosh

Comment (etorreborre) [inline src/Icicle/Data/DateTime.hs]: Can ambiata/delorean be used here?

Comment (amosr) [inline src/Icicle/Data/DateTime.hs]: sounds ideal. I'm not sure. I don't think delorean is open source right now either

File: src/Icicle/Data/DateTime.hs [hunk 2]
```diff
@@ -31,11 +33,49 @@ renderDate
...
+
+dateOfYMD :: Int -> Int -> Int -> Maybe DateTime
+dateOfYMD y m d
+ | isValid y m d =
+   Just $ unsafeDateOfYMD y m d
+ | otherwise =
+   Nothing
+
+isValid :: Int -> Int -> Int -> Bool
+isValid year month day =
```
Comment (amosr) [inline src/Icicle/Data/DateTime.hs]: http://hackage.haskell.org/package/time-1.5.0.1/docs/Data-Time-Calendar.html#v:fromGregorianValid
?

Comment (HuwCampbell) [inline src/Icicle/Data/DateTime.hs]: Sure, ok.

**Comment by amosr** [comment]
:+1: looks good

**Comment by HuwCampbell** [comment]
Cool, I was just about to push another bit which does Date Values as first class. I can do it separately though.

**Comment by HuwCampbell** [comment]
I don't know, it seems to make sense here.

**Comment by amosr** [comment]
yeah! looks good to me

**Pull request merged**